flang/lib/Lower/OpenMP.cpp

```diff
@@ -2400,6 +2417,76 @@ genEnterExitDataOp(Fortran::lower::AbstractConverter &converter,
                                    deviceOperand, nowaitAttr, mapOperands);
 }
 
+static void genBodyOfTargetOp(
+    Fortran::lower::AbstractConverter &converter, mlir::omp::TargetOp &targetOp,
+    const llvm::SmallVector<mlir::Type> &mapSymTypes,
+    const llvm::SmallVector<mlir::Location> &mapSymLocs,
+    const llvm::SmallVector<const Fortran::semantics::Symbol *> &mapSymbols,
+    const mlir::Location &currentLocation) {
+  assert(mapSymTypes.size() == mapSymLocs.size());
+
+  fir::FirOpBuilder &firOpBuilder = converter.getFirOpBuilder();
+  mlir::Region &region = targetOp.getRegion();
+
+  firOpBuilder.createBlock(&region, {}, mapSymTypes, mapSymLocs);
+  firOpBuilder.create<mlir::omp::TerminatorOp>(currentLocation);
+  firOpBuilder.setInsertionPointToStart(&region.front());
+
+  unsigned argIndex = 0;
+  unsigned blockArgsIndex = mapSymbols.size();
+
+  auto extractBoundArgs = [&](auto n) {
+    llvm::SmallVector<mlir::Value> argExtents;
+    while (n--) {
+      argExtents.push_back(fir::getBase(region.getArgument(blockArgsIndex)));
+      blockArgsIndex++;
+    }
+    return argExtents;
+  };
+
+  for (const Fortran::semantics::Symbol *sym : mapSymbols) {
+    const mlir::BlockArgument &arg = region.getArgument(argIndex);
+    fir::ExtendedValue extVal = converter.getSymbolExtendedValue(*sym);
+    mlir::Value val = fir::getBase(arg);
```

Comment:
@kiranchandramohan I am getting the following error for CharBoxTypes on this statement. Any idea why?

`character buffer should be in CharBoxValue`

Reply:
What is the type of `arg` here? Is the intention to call getBase on an mlir::Value or an `extendedValue`?

Reply:
Hi, I thought we need to bind the symbol to the `fir::getBase(arg)` but we can directly bind it to the arg.

All the cases in the new `flang/test/Lower/OpenMP/map-types-and-sizes.f90` test are now passing.

@razvanlupusoru Please let me know if you think the test itself needs any corrections.

Reply:
Does that mean you do not see the error any more?

Or did you intend the second part of the comment for `map_type here for Fortran pointers`?

Reply:
This error was only for using `getBase(arg)` when arg was of `CharBoxType`. With the change the error is gone and the Character test passes, along with other tests.

This wasn't related to the `map_type` for `Fortran pointers` issue.

Reply:
The region arguments were set using the type of the `baseOp` which was captured previously using `converter.getSymbolAddress`. I guess you cannot do `getBase` of such values.
```
      mlir::Value baseOp = converter.getSymbolAddress(sym);
```
It will be good to get this portion reviewed by @jeanPerier.

Reply:
The update to use getSymbolExtendedValue and the fir::ExtendedValue and looks good to me.